File: Part1._Enumeration_problem/Lesson2._Generation_method/Combination.c
#include <stdio.h>

int main()
{
    FILE *fi = NULL, *fo = NULL;
    int i = 0;

    fi = fopen("SUBSET.INP", "r");
    if (fi != NULL)
    {
        int n = 0, k = 0;
        fscanf(fi, "%d %d", &n, &k);
        fclose(fi);
        if (1 <= k && k <= n && n <= 30)
        {
            fo = fopen("SUBSET.OUT", "w");
            int x[n + 1];

            // Khởi tạo cấu hình ban đầu
            for (i = 1; i <= k; i++)
            {
                x[i] = i;
            }

            do
            {
                // Xuất cấu hình hiện tại ra file
                fprintf(fo, "{");
                for (i = 1; i < k; i++)
                {
                    fprintf(fo, "%d, ", x[i]);
                }
                fprintf(fo, "%d}\n", x[k]);

                // Sinh cấu hình kế tiếp
                i = k;
                while (i > 0 && x[i] == n - k + i)
                {
                    i--;
                }
                if (i > 0) // Khi chưa gặp cấu hình cuối cùng
                {
                    x[i] = x[i] + 1;
                    for (int j = i + 1; j <= k; j++)
                    {
                        x[j] = x[j - 1] + 1;
                    }
                }
            } while (i != 0); // Đã gặp cấu hình cuối cùng
            fclose(fo);
        }
    }
    return 0;
}

File: Part1._Enumeration_problem/Lesson2._Generation_method/Combination.pas
program Combination;
const
    InputFile = 'SUBSET.INP';
    OutputFile = 'SUBSET.OUT';
    max = 30;
var
    x: array[1..max] of Integer;
    n, k, i, j: Integer;
    f: Text;
begin
    Assign(f, InputFile); Reset(f);
    ReadLn(f, n, k);
    Close(f);
    Assign(f, OutputFile); Rewrite(f);
    for i := 1 to k do x[i] := i; {x := 12…k (Cấu hình khởi tạo)}
    repeat
        {In ra cấu hình hiện tại}
        Write(f, '{');
        for i := 1 to k - 1 do Write(f, x[i], ', ');
        WriteLn(f, x[k], '}');
        {Sinh tiếp}
        i := k; {Xét từ cuối dãy lên tìm x[i] chưa đạt giới hạn trên n - k + i}
        while (i > 0) and (x[i] = n - k + i) do Dec(i);
        if i > 0 then {Nếu chưa lùi đến 0 có nghĩa là chưa phải cấu hình kết thúc}
            begin
                Inc(x[i]); {Tăng x[i] lên 1, Đặt các phần tử đứng sau x[i] bằng giới hạn dưới của nó}
                for j := i + 1 to k do x[j] := x[j - 1] + 1;
            end;
    until i = 0; {Lùi đến tận 0 có nghĩa là tất cả các phần tử đã đạt giới hạn trên - hết cấu hình}
    Close(f);
end.